### PR TITLE
[LV]: Teach LV to recursively (de)interleave. 

### PR DESCRIPTION
**Pull request by hassnaaHamdi**
Currently available intrinsics are only ld2/st2, which don't support interleaving factor > 2.
This patch teaches the LV to use ld2/st2 recursively to support high interleaving factors.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 1]
```diff
@@ -2519,7 +2533,7 @@ void InnerLoopVectorizer::vectorizeInterleaveGroup(
                              unsigned Part, Value *MaskForGaps) -> Value * {
     if (VF.isScalable()) {
       assert(!MaskForGaps && "Interleaved groups with gaps are not supported.");
-      assert(InterleaveFactor == 2 &&
+      assert(isPowerOf2_32(InterleaveFactor)  &&
              "Unsupported deinterleave factor for scalable vectors");
       auto *BlockInMaskPart = State.get(BlockInMask, Part);
       SmallVector<Value *, 2> Ops = {BlockInMaskPart, BlockInMaskPart};
```
Comment (Mel-Chen) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:2539]: The mask of masked interleaved accesses also requires an interleave tree to generate the correct mask.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 2]
```diff
@@ -2572,23 +2586,40 @@ void InnerLoopVectorizer::vectorizeInterleaveGroup(
...
-            Intrinsic::experimental_vector_deinterleave2, VecTy, NewLoads[Part],
-            /*FMFSource=*/nullptr, "strided.vec");
+        
+        std::queue<Value *>Queue;
+        Queue.push(NewLoads[Part]);
+        // NonLeaf represents how many times we will do deinterleaving,
+        // think of it as a tree, each node will be deinterleaved, untill we reach to
+        // the leaf nodes which will be the final results of deinterleaving.
+        unsigned NonLeaf = InterleaveFactor - 1;
+        for (unsigned i = 0; i < NonLeaf; i ++) {
```
Comment (Mel-Chen) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:2601]: `i` --> `I`
`i ++` --> `I++`

**Comment by fhahn** [comment]
It loos like at the TargetLowering, LV and InterleavedAccessPass changes could be decoupled?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 3]
```diff
@@ -2572,23 +2586,40 @@ void InnerLoopVectorizer::vectorizeInterleaveGroup(
...
-        for (unsigned I = 0; I < InterleaveFactor; ++I) {
+        for (unsigned I = 0; I < InterleaveFactor && !Queue.empty(); ++I) {
           Instruction *Member = Group->getMember(I);
 
           if (!Member)
             continue;
 
-          Value *StridedVec = Builder.CreateExtractValue(DI, I);
+          auto StridedVec = Queue.front();
+          Queue.pop();
```
Comment (Mel-Chen) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:2622]: Here is a example:
A vector 0 1 2 3 4 5 6 7
If we do deinterleave 4 on the vector, we should get:
member 0: 0 4
member 1: 1 5
member 2: 2 6
member 3: 3 7
But the Queue in your change may like:  0 4, 2 6, 1 5, 3 7.
Please confirm the Queue is sorted by a correct rank.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 4]
```diff
@@ -2681,6 +2712,7 @@ void InnerLoopVectorizer::vectorizeInterleaveGroup(
 
     // Interleave all the smaller vectors into one wider vector.
     Value *IVec = interleaveVectors(Builder, StoredVecs, "interleaved.vec");
+    //LLVM_DEBUG(dbgs() << "interleaved vec: "; IVec->dump());
```
Comment (Mel-Chen) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:2715]: Please remove it.

**Comment by paulwalker-arm** [comment]
> It loos like at the TargetLowering, LV and InterleavedAccessPass changes could be decoupled?

I agree.  Please can you split this PR into two PRs with the first only concerned with the correct lowering of the new emulated ld4/st4 IR sequence.  The second PR then teaches LoopVectorize how to use them.

**Comment by davemgreen** [comment]
Hi - Is there a plan for how to handle ld3? We have seen a lot of issues recently with the canonical shuffle representation for fixed-vector ld2/ld3/ld4, and I was wondering if it made sense to move away from shuffles for fixed-length too.

**Comment by paulwalker-arm** [comment]
> Hi - Is there a plan for how to handle ld3? We have seen a lot of issues recently with the canonical shuffle representation for fixed-vector ld2/ld3/ld4, and I was wondering if it made sense to move away from shuffles for fixed-length too.

There is but that will require a new intrinsic.  My hope is that rather than having an intrinsic per interleave factor we could model them all using interleave2 and interleave3 (once it's created).  This is why we've started with ld4/st4 support to see if there are any pitfalls to this approach.

Personally I'd love us to move to using these intrinsics for all vector types because it will streamline several code paths.

**Comment by efriedma-quic** [comment]
Doing deinterleaving as trees sort of makes sense for high interleaving factors... I've seen loops that benefit from deinterleaving with interleave factors as high as 12.  I'm a little concerned the abstraction layers here are going to make cost modeling less accurate, though; ideally, the vectorizer should be able to estimate the cost of an ld4.

**Comment by topperc** [comment]
> > Hi - Is there a plan for how to handle ld3? We have seen a lot of issues recently with the canonical shuffle representation for fixed-vector ld2/ld3/ld4, and I was wondering if it made sense to move away from shuffles for fixed-length too.
> 
> There is but that will require a new intrinsic. My hope is that rather than having an intrinsic per interleave factor we could model them all using interleave2 and interleave3 (once it's created). This is why we've started with ld4/st4 support to see if there are any pitfalls to this approach.
> 
> Personally I'd love us to move to using these intrinsics for all vector types because it will streamline several code paths.

RISC-V has interleave loads for up to 8. So I guess we would need interleave5 and interleave7?

**Comment by paulwalker-arm** [comment]
> RISC-V has interleave loads for up to 8. So I guess we would need interleave5 and interleave7?

Yes, sorry.  I guess I meant "Hopefully we can emulate all required interleave factors by only implement specific intrinsics for factors that are a prime number"?  An alternative proposal is to have intrinsics for all but then lower them to sequences of fewer intrinsics within the InterleavedAccess pass or perhaps even SelectionDAGBuilder.  I suppose this really depends on how awkward cost modelling the sequences turns out to be.

@efriedma-quic - Is your concern related to vectorisation or the costing of already vectorised code?

**Comment by efriedma-quic** [comment]
Given the way the pass pipeline is structured, cost modeling in the vectorizer itself tends to be more important than modeling in subsequent passes.  I guess maybe it's not a big deal what the vectorizer generates if the vectorizer itself has some way to get the correct numbers.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 1 above)
Comment (hassnaaHamdi) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:2539]: Could you please give an example for a case that uses masked interleaved accesses ?
I have commented the code of creating masked load (the call to CreateGroupMask lambda function), and reran the tests but all tests ran successfully. It seems that for the interleaved accesses all the loads are aligned not masked.

**Comment by davemgreen** [comment]
The loop vectorizer will produce costs via getInterleavedMemoryOpCost so should be fine as far as I understand. If there are no combines later on (either uncosted in instcombine or costed in vector-combine) that work with vector.interleave/vector.deinterleave then they can break the canonical patterns that the backend is expecting to generate ld2/ld4 from. I'm hoping that if we can move to interleave/deinterleave, that should fix some of the problems we have at the moment.

I have recently been adding costs for the existing shuffles we find for fixed length vectors, in an attempt to reduce the number of times we break apart the load+shuffle (or store+shuffle), and have to either attempt to repair it or fall back to worse generation in the backend. I would say that in general costing for single-instructions is fine, two instructions making a pattern (like shuffle(load) or store(shuffle)) are do-able but start to get unreliable, and three-instruction plus becomes difficult to cost well.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 5]
```diff
@@ -2126,10 +2127,39 @@ static Value *interleaveVectors(IRBuilderBase &Builder, ArrayRef<Value *> Vals,
...
+    for (unsigned I = 0, J = InterleaveFactor / 2, K = 0; K < InterleaveFactor;
+         K++) {
+      if (K % 2 == 0) {
+        InterleavingValues[K] = Vals[I];
+        I++;
+      } else {
+        InterleavingValues[K] = Vals[J];
+        J++;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (hassnaaHamdi) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2146]: Yes, my intent is to make it generic.
I think to make it generic, There will be multiple sorting during the interleave/deinterleave, not only at the end. correct ?

[the remark this answers is not in the document]
Comment (paulwalker-arm) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2146]: Yes I believe so.  It'll be a continuous process of "ordering the operands and then interleaving them" until you have only one vector (or continuous "deinterleave and then order the results" until you have the required N vectors).

**Comment by Mel-Chen** [comment]
Would it be possible to directly add (de)interleave intrinsic 4 to achieve this? Such an implementation should be simpler and more maintainable. 
Using prime factor (de)interleave intrinsics to combine non-prime interleave factors would only be effective in saving the number of intrinsics needed when supporting large interleave factors. 
For RISCV, the current largest factor is 8. (Not sure if other targets support larger factors.)

**Comment by paulwalker-arm** [comment]
> Would it be possible to directly add (de)interleave intrinsic 4 to achieve this? 

A significant risk of not being able to identify the larger interleave factors would be the main reason to introduce dedicated intrinsics. However, I'd expect the IR to emulate an 8-way interleave to be pretty fixed so I'd rather wait to see if this is proved incorrect before going down that route.

At the end of the day the code this PR will introduce will be required anyway to lower (de)interleave intrinsics a target does not support, so there shouldn't be much wasted effort.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (hassnaaHamdi) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Done

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 6]
```diff
@@ -2126,10 +2127,39 @@ static Value *interleaveVectors(IRBuilderBase &Builder, ArrayRef<Value *> Vals,
   // Scalable vectors cannot use arbitrary shufflevectors (only splats), so
   // must use intrinsics to interleave.
   if (VecTy->isScalableTy()) {
-    VectorType *WideVecTy = VectorType::getDoubleElementsVectorType(VecTy);
-    return Builder.CreateIntrinsic(WideVecTy, Intrinsic::vector_interleave2,
-                                   Vals,
-                                   /*FMFSource=*/nullptr, Name);
+    unsigned InterleaveFactor = Vals.size();
```
[the remark this answers is not in the document]
Comment (hassnaaHamdi) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2136]: Done

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 7]
```diff
@@ -2126,10 +2127,39 @@ static Value *interleaveVectors(IRBuilderBase &Builder, ArrayRef<Value *> Vals,
...
+        I++;
+      } else {
+        InterleavingValues[K] = Vals[J];
+        J++;
+      }
+    }
+#ifndef NDEBUG
+    for (Value *Val : InterleavingValues)
+      assert(Val && "NULL Interleaving Value");
+#endif
```
[the remark this answers is not in the document]
Comment (hassnaaHamdi) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Done

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 8]
```diff
@@ -2291,23 +2329,69 @@ void VPInterleaveRecipe::execute(VPTransformState &State) {
...
+          unsigned NumRemainingValues = TempDeinterleavedValues.size() - I - 1;
+          if (NumRemainingValues > 2 && isPowerOf2_32(NumRemainingValues)) {
+            // these remaining values represent a new level in the tree,
+            // Reorder the values to match the correct deinterleaving order.
+            std::vector<Value *> RemainingValues(
+                TempDeinterleavedValues.begin() + I + 1,
+                TempDeinterleavedValues.end());
+            unsigned Middle = NumRemainingValues / 2;
+            for (unsigned J = 0, K = I + 1; J < NumRemainingValues;
+                 J += 2, K++) {
```
[the remark this answers is not in the document]
Comment (hassnaaHamdi) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: done

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 9]
```diff
@@ -2910,22 +2919,48 @@ void VPInterleaveRecipe::execute(VPTransformState &State) {
...
+      // so must use intrinsics to deinterleave.
+
+      SmallVector<Value *> DeinterleavedValues(InterleaveFactor);
+      DeinterleavedValues[0] = NewLoad;
+      // For the case of InterleaveFactor > 2, we will have to do recursive
+      // deinterleaving, because the current available deinterleave intrinsic
+      // supports only Factor of 2, otherwise it will bailout after first
+      // iteration.
+      // As we are deinterleaving, the values will be doubled until reachingt
+      // to the InterleaveFactor.
```
Comment (paulwalker-arm) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2935]:
```suggestion
      // When deinterleaving, the number of values will double until we have "InterleaveFactor".
```

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 10]
```diff
@@ -2862,15 +2874,12 @@ void VPInterleaveRecipe::execute(VPTransformState &State) {
...
              "Unsupported deinterleave factor for scalable vectors");
       auto *ResBlockInMask = State.get(BlockInMask);
-      SmallVector<Value *, 2> Ops = {ResBlockInMask, ResBlockInMask};
-      auto *MaskTy = VectorType::get(State.Builder.getInt1Ty(),
-                                     State.VF.getKnownMinValue() * 2, true);
-      return State.Builder.CreateIntrinsic(
-          MaskTy, Intrinsic::vector_interleave2, Ops,
-          /*FMFSource=*/nullptr, "interleaved.mask");
+      SmallVector<Value *> Ops;
+      Ops.resize(InterleaveFactor, ResBlockInMask);
```
Comment (paulwalker-arm) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2881]: Not 100% sure but does `SmallVector<Value *> Ops(InterleaveFactor, ResBlockInMask);` do the same thing?

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 11]
```diff
@@ -2910,22 +2919,48 @@ void VPInterleaveRecipe::execute(VPTransformState &State) {
...
+            DeinterleavedValues[NumVectors * I + J] =
+                State.Builder.CreateExtractValue(TempDeinterleavedValues[J], I);
+      }
 
-        if (!Member)
+#ifndef NDEBUG
+      for (Value *Val : DeinterleavedValues)
+        assert(Val && "NULL Deinterleaved Value");
+#endif
+      for (unsigned I = 0, J = 0; I < InterleaveFactor; ++I) {
```
Comment (paulwalker-arm) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2956]: `J` is unused?

**Review by paulwalker-arm — COMMENTED** [review]
Very nice.  You just need to sort out the testing and the PR will be good to go.

File: llvm/test/Transforms/InterleavedAccess/AArch64/sve-deinterleave4.ll [hunk 12]
```diff
@@ -1,5 +1,5 @@
 ; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 2
-; RUN: opt < %s -passes=interleaved-access -mtriple=aarch64-linux-gnu -mattr=+sve -S | FileCheck %s
+; RUN: opt < %s -passes=loop-vectorize,interleaved-access -mtriple=aarch64-linux-gnu -mattr=+sve -S | FileCheck %s
```
[the remark this answers is not in the document]
Comment (hassnaaHamdi) [reply llvm/test/Transforms/InterleavedAccess/AArch64/sve-deinterleave4.ll:2]: I understand the reason behind first suggestion is that generating the final intrinsics requires applying multiple passes (LoopVectorize and InterleavedAccess) and the destination of Transforms/PhaseOrdering represents that case, correct ?

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 11 above)
Comment (hassnaaHamdi) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2956]: It's used at the end of the loop for updating the final state/(the used interleaved values)

File: llvm/test/Transforms/InterleavedAccess/AArch64/sve-deinterleave4.ll [hunk 13]
```diff
@@ -136,3 +136,4 @@ define void @negative_deinterleave4_test(ptr %src) {
 
   ret void
 }
+
```
Comment (paulwalker-arm) [inline llvm/test/Transforms/InterleavedAccess/AArch64/sve-deinterleave4.ll:139]: Please revert this unnecessary change.

File: llvm/test/Transforms/InterleavedAccess/AArch64/sve-deinterleave4.ll (hunk 12 above)
[the remark this answers is not in the document]
Comment (paulwalker-arm) [reply llvm/test/Transforms/InterleavedAccess/AArch64/sve-deinterleave4.ll]: Yes that's correct. Thanks for making the change.

**Review by paulwalker-arm — APPROVED** [review]

**Comment by hassnaaHamdi** [comment]
I'm going to rebase and land the patch in the next days if there are no further comments.

**Review by Mel-Chen — COMMENTED** [review]
Thanks for your contribution. Here is some questions for this patch:
1. I believe we requires InterleavedAccessPass(IAP) support before this patch. Does IAP now have the ability to convert the power of 2 factor?
2. Do you have lit test cases for masked interleaved accesses and reverse interleaved accesses with power of 2 factor?
3. Last question is for the future work. RISCV supports factor 6. If we have interleave3/deinterleave3 intrinsics, can the approach in this patch support factor 6? Or will it require a lot of modifications to support factor 6?

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 14]
```diff
@@ -2779,10 +2779,21 @@ static Value *interleaveVectors(IRBuilderBase &Builder, ArrayRef<Value *> Vals,
...
-    return Builder.CreateIntrinsic(WideVecTy, Intrinsic::vector_interleave2,
-                                   Vals,
-                                   /*FMFSource=*/nullptr, Name);
+    unsigned InterleaveFactor = Vals.size();
+    SmallVector<Value *> InterleavingValues(Vals);
+    // When interleaving, the number of values will be shrunk until we have the
+    // single final interleaved value.
+    VectorType *InterleaveTy =
+        cast<VectorType>(InterleavingValues[0]->getType());
+    for (unsigned Midpoint = Factor / 2; Midpoint > 0; Midpoint /= 2) {
```
[the remark this answers is not in the document]
Comment (hassnaaHamdi) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2788]: Hi @Mel-Chen 
Thanks for looking at the patch.
The assert statement is already added before calling the interleaveVectors(..) function.

About your questions above:
1. yes I have landed a patch for adding support to the InterleaveAccessPass to support reading the (de)interleave tree pattern.
2. Going to add them.
3. If we have (de)interleave3 intrinsics, then we will have to do same logic for recursive (de)interleave3, and then the extra needed work will be representing the interleave factor by multiple of 2 and 3. so for the case of factor 6, we will do single iteration of (de)interleave2 then single iteration of (de)interleave3. The same logic will be applied for all factors that consist of multiples of 2 and 3 only.

[the remark this answers is not in the document]
Comment (hassnaaHamdi) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2786]: Hi @Mel-Chen 
Are you satisfied about the latest changes ?

File: llvm/test/Transforms/InterleavedAccess/AArch64/sve-deinterleave4.ll (hunk 13 above)
Comment (Mel-Chen) [reply llvm/test/Transforms/InterleavedAccess/AArch64/sve-deinterleave4.ll:139]: Please remove this blank line.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 15]
```diff
@@ -2916,22 +2921,49 @@ void VPInterleaveRecipe::execute(VPTransformState &State) {
...
-        // so must use intrinsics to deinterleave.
-      Value *DI = State.Builder.CreateIntrinsic(
-          Intrinsic::vector_deinterleave2, VecTy, NewLoad,
-          /*FMFSource=*/nullptr, "strided.vec");
-      unsigned J = 0;
-      for (unsigned I = 0; I < InterleaveFactor; ++I) {
-        Instruction *Member = Group->getMember(I);
+      // Scalable vectors cannot use arbitrary shufflevectors (only splats),
+      // so must use intrinsics to deinterleave.
+
```
Comment (Mel-Chen) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2929]: Remove this blank line.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (Mel-Chen) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2786]: > The assert statement is already added before calling the interleaveVectors(..) function.

I think we still need an assert before the for loop, as this is a standalone function. This will ensure that no caller inadvertently passes an invalid factor in the future.

**Comment by hassnaaHamdi** [comment]
Hi @Mel-Chen 
Thanks for your review.
I have resolved your comments, I think the patch is ready now to be landed ?

**Review by Mel-Chen — APPROVED** [review]
The IR looks correct.
I temporarily accept this solution, although it will generate Σ(2^i), i = 0 to log2(factor)-1 interleave2/deinterleave2. 
Please wait for other reviewers @fhahn @ayalz for a few days.

**Comment by omjavaid** [comment]
This PR appears to be breaking LLVM build on AArch64 SVE Linux buildbots
https://lab.llvm.org/buildbot/#/builders/143/builds/4462
https://lab.llvm.org/buildbot/#/builders/17/builds/4902
https://lab.llvm.org/buildbot/#/builders/4/builds/4399
https://lab.llvm.org/buildbot/#/builders/41/builds/4299

**Comment by omjavaid** [comment]
I have reverted this change temporarily to fix the buildbots. Please review the change. Thanks!
